src/scanner.c

```diff
@@ -0,0 +1,431 @@
+  }
+
+#define stop()                       \
+  {                                  \
+    print("stop %s\n", str(peek())); \
+    lexer->mark_end(lexer);          \
+  }
+
+#define set(symbol)                        \
+  {                                        \
+    print("set %s\n", TokenTypes[symbol]); \
+    lexer->result_symbol = symbol;         \
+  }
+
+#define ret(function, result)                                 \
+  print("%s() -> %s\n", function, result ? "true" : "false"); \
+  return result;
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+#define STRING_RESIZE(vec, _cap)                                                           \
+  (vec)->data = realloc((vec)->data, (_cap + 1) * sizeof((vec)->data[0]));                 \
+  assert((vec)->data != NULL);                                                             \
+  memset((vec)->data + (vec)->len, 0, ((_cap + 1) - (vec)->len) * sizeof((vec)->data[0])); \
+  (vec)->cap = (_cap);
+
+#define STRING_GROW(vec, _cap)    \
+  if ((vec)->cap < (_cap)) {      \
+    STRING_RESIZE((vec), (_cap)); \
+  }
+
+#define STRING_PUSH(vec, el)                       \
+  if ((vec)->cap == (vec)->len) {                  \
+    STRING_RESIZE((vec), MAX(16, (vec)->len * 2)); \
+  }                                                \
+  (vec)->data[(vec)->len++] = (el);
+
+#define STRING_INIT(vec)                        \
+  {                                             \
+    (vec) = calloc(1, sizeof(String));          \
+    (vec)->data = calloc(1, sizeof(char) * 16); \
+  }
+
+#define STRING_FREE(vec)                        \
+  {                                             \
+    if ((vec)->data != NULL) free((vec)->data); \
+    free((vec));                                \
+  }
+
+#define STRING_CLEAR(vec)                              \
+  {                                                    \
+    (vec)->len = 0;                                    \
+    memset((vec)->data, 0, (vec)->cap * sizeof(char)); \
+  }
+
+enum TokenType {
+  HEREDOC_START,
+  HEREDOC_START_NEWLINE,
+  HEREDOC_BODY,
+  HEREDOC_END_NEWLINE,
+  HEREDOC_END,
+  EMBEDDED_OPENING_BRACE,
+};
+
+const char *TokenTypes[] = {
+    "HEREDOC_START",           //
+    "HEREDOC_START_NEWLINE",   //
+    "HEREDOC_BODY",            //
+    "HEREDOC_END_NEWLINE",     //
+    "HEREDOC_END",             //
+    "EMBEDDED_OPENING_BRACE",  //
+};
+
+typedef struct {
+  uint32_t cap;
+  uint32_t len;
+  char *data;
+} String;
+
+typedef struct {
+  String *delimiter;
+  bool is_nowdoc;
+  bool did_start;
+  bool did_end;
+} Scanner;
+
+static char *str(int32_t chr) {
+  switch (chr) {
+    case '\n':
+      return "\\n";
+    case '\r':
+      return "\\r";
+    case '\t':
+      return "\\t";
+    case ' ':
+      return "\\s";
+    case '\0':
+      return "\\0";
+    default:
+      if (iswspace(chr)) {
+        return "\\s";
+      }
+
+      char *str = calloc(2, sizeof(char));
```

Comment:
fixed, I didnt have debug set to 1 to test leaks when i shouldve...

Reply:
thanks, this works but I don't think it is the greatest, since it uses static variables and thus might not be able to be optimized away (e.g. if debug=0)
there is history here, this code had bugs in the c++ version too #37 which got fixed by always allocating the return value. I think this is fine, since it should be optimized away, so don't shy away from allocation

if you really want to avoid allocation from this function, I think you should return a char array that is stack allocated. Here's an example of how this can be done
```C
struct x {
    char str[3];
};
struct x escape(char c) {
    struct x ret = {.str = {c,0,0}};
    switch(c) {
    case '\n':
        ret.str[1]='n';
    case '\r':
        ret.str[1]='r';
    case '\t':
        ret.str[1]='t';
    case ' ':
        ret.str[1]='s';
    case '\0':
        ret.str[1]='0';
    default: 
        return ret;
    }
    ret.str[0]='\\';
    return ret;
}
```

Reply:
Yes but it does not generate the same error, the linked issue returned a const char pointer, this returns a static char array and clang nor the fuzzer complain..so I think it's ok. I can't comment on the optimization though

Reply:
check out godbolt: https://godbolt.org/z/46c4eqr1e
even with `-O3`, any call to `str` cannot be optimized away.
if you delete the modifications to `buffer`, then `str` calls can be optimized away. This is because it is modifying global memory, and could have arbitrary side effects.

we want `str` calls to be optimized away, because it is only used in debug mode.
Thus, I would prefer either returning a static buffer (as given), allocate a string or pass a buffer into this function.

Here's an example which allocates a string: https://godbolt.org/z/ah1qaP6EY 
as you can see, the call is also optimized away in `-O3`.

Reply:
wow, thanks for the examples!

should be fixed now